sklearn/semi_supervised/self_training.py

```diff
@@ -0,0 +1,239 @@
+        Should be in [0, 1).
+
+    max_iter : integer
+        Maximum number of iterations allowed. Should be greater than or equal
+        to 0.
+
+    Attributes
+    ----------
+    base_estimator_: estimator object
+        The fitted estimator.
+
+    y_labeled_ : array, shape = (n_samples)
+        The labels assigned to unlabeled datapoints during fitting.
+
+    X_labeled_ : array, shape = (n_samples, n_features)
+        The labeled samples used for the final fit.
+
+    y_labeled_iter_ : array, shape = (n_samples)
+        The iteration in which each sample was labeled. When a sample has
+        iteration 0, the sample was labeled in the given dataset.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn import datasets
+    >>> from sklearn.semi_supervised import SelfTrainingClassifier
+    >>> from sklearn.svm import SVC
+    >>> svc = SVC(probability=True, gamma="auto")
+    >>> self_training_model = SelfTrainingClassifier(svc)
+    >>> iris = datasets.load_iris()
+    >>> rng = np.random.RandomState(42)
+    >>> random_unlabeled_points = rng.rand(len(iris.target)) < 0.3
+    >>> labels = np.copy(iris.target)
+    >>> labels[random_unlabeled_points] = -1
+    >>> self_training_model.fit(iris.data, labels)
+    ... # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    SelfTrainingClassifier(...)
+
+    References
+    ----------
+    David Yarowsky. 1995. Unsupervised word sense disambiguation rivaling
+    supervised methods. In Proceedings of the 33rd annual meeting on
+    Association for Computational Linguistics (ACL '95). Association for
+    Computational Linguistics, Stroudsburg, PA, USA, 189-196. DOI:
+    https://doi.org/10.3115/981658.981684
+    """
+    def __init__(self,
+                 base_estimator,
+                 threshold=0.75,
+                 max_iter=100):
+        self.base_estimator = base_estimator
+        self.threshold = threshold
+        self.max_iter = max_iter
+
+    def fit(self, X, y):
+        """
+        Fits this ``SelfTrainingClassifier`` to dataset, using the
+        base_estimator passed. The fitted base_estimator is stored as
+        ``self.base_estimator_``.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            array representing the data
+
+        y : array-like, shape = (n_samples, 1)
+            array representing the labels
+
+        Returns
+        -------
+        self: returns an instance of self.
+        """
+        X, y = check_X_y(X, y)
+        _check_estimator(self.base_estimator)
+        self.base_estimator_ = clone(self.base_estimator)
+
+        if not 0 <= self.max_iter:
+            raise ValueError("max_iter must be >= 0")
+
+        if not 0 <= self.threshold < 1:
+            raise ValueError("threshold must be in [0,1)")
+
+        # Data usable for supervised training
+        self.X_labeled_ = X[safe_mask(X, np.where(y != -1))][0]
```

Comment:
I don't think you need np.where here.

Reply:
Why is that? Do you suggest a different method for performance reasons?

Reply:
Yeah, `X[safe_mask(X, y != -1)]` would have worked too.